app/controllers/invitations_controller.rb

```diff
@@ -0,0 +1,87 @@
+class InvitationsController < ApplicationController
+  before_action :restrict_access_on_create, only: [ :new, :create ]
+
+  def new
+    @invitation = Invitation.new
+  end
+
+  def create
+    @invitation = Invitation.new(invitation_params)
+    invitation_token = SecureRandom.urlsafe_base64
+    @invitation.digest = digest invitation_token
+
+    if (@invitation.save)
+      # Just for the test purposes.
+      puts edit_invitation_url(invitation_token, email: @invitation.email)
+      @token = invitation_token
+      @email = @invitation.email
+      render file: "user_mailer/invitation.text.erb"
+
+      # Send mail
+      UserMailer.invitation(invitation_token, @invitation.email).deliver
+    else
+      render "new"
+    end
+  end
+
+  def edit
+    @invitation = Invitation.find_by(email: params[:email])
+
+    unless (@invitation && BCrypt::Password.new(@invitation.digest).is_password?(params[:id]))
+      return redirect_to root_path, alert: "Invalid activation link!"
+    end
+
+    if user_signed_in?
+      current_user.update(user_type: :teacher) if current_user.email == params[:email]
+      return redirect_to root_path
+    end
+
+    @user = User.new(email: params[:email])
+    render_registration
+  end
+
+  def update
+    @user = User.new(activation_params)
+    @user.user_type = :teacher
+
+    if @user.save
+      sign_in(:user, @user)
+      redirect_to root_path
+    else
+      render_registration
+    end
+  end
+
+private
+  # Only teachers and administrators have permission to invite new users 
+  def restrict_access_on_create
+    return redirect_to new_user_session_path, alert: "You should be logged in." unless user_signed_in?
```

Comment:
There's an `authenticate_user!` method in devise. Example usage in the [`TasksController`](https://github.com/auduchinok/hwproj/blob/master/app/controllers/tasks_controller.rb#L8).